src/main/cpp/src/zorder.cu

```diff
@@ -27,6 +27,105 @@
 #include <thrust/iterator/constant_iterator.h>
 #include <thrust/iterator/counting_iterator.h>
 
+namespace {
+
+// pretends to be an array of uint32_t, but really only stores
+// the data in a long with a set number of bits allocated for
+// each item (num_bits_per_entry)
+struct long_backed_array {
+  using data_type = uint64_t;
+  long_backed_array() = delete;
+  __device__ explicit long_backed_array(int32_t num_bits_per_entry): data(0),
+    num_bits_per_entry(num_bits_per_entry),  mask(static_cast<data_type>((1L << num_bits_per_entry) - 1)) {}
+
+  __device__ uint32_t operator[](int32_t i) const {
+    int32_t offset = num_bits_per_entry * i;
+    return (data >> offset) & mask;
+  }
+
+  __device__ void set(int32_t i, uint32_t value) {
+    int32_t offset = i * num_bits_per_entry;
+    data_type masked_data = data & ~(static_cast<data_type>(mask) << offset);
+    data = masked_data | (static_cast<data_type>(value & mask) << offset);
+  }
+
+private:
+  uint64_t data;
+  int32_t const num_bits_per_entry;
+  uint32_t const mask;
+};
+
+
+// Most of the hilbert index code is based off of the work done by David Moten at
+// https://github.com/davidmoten/hilbert-curve, which has the following Note in
+// the code too
+// This algorithm is derived from work done by John Skilling and published
+// in "Programming the Hilbert curve". (c) 2004 American Institute of Physics.
+// With thanks also to Paul Chernoch who published a C# algorithm for Skilling's
+// work on StackOverflow and
+// <a href="https://github.com/paulchernoch/HilbertTransformation">GitHub</a>.
+__device__ uint64_t to_hilbert_index(long_backed_array const & transposed_index,
+        int32_t const num_bits_per_entry, int32_t const num_dimensions) {
+  uint64_t b = 0;
```

Comment:
these are sized to match the `long_backed_array`, correct? I would expect these to be
`long_backed_array::data_type` instead of `uint64_t`. It is a bit pedantic, but if the backing store changes, it would be easy to miss changing these.

Reply:
They are but also they are not. I picked the size of long_backed_array because of the size of the output needed in this case. So for me you are pushing for the dependency to go in the wrong direction. The size of long_backed_array is set by the desired output size, not the other way around.  Perhaps I should just change it to be templated. But I don't really want to because what we are trying to replace does not have more options than a long.